src/opencl/androidbackup_kernel.cl

```diff
@@ -0,0 +1,57 @@
+/*
+ * This software is Copyright (c) 2018 Dhiru Kholia, Copyright (c) 2018 magnum
+ * and it is hereby released to the general public under the following terms:
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ */
+
+#include "pbkdf2_hmac_sha1_kernel.cl"
+#define AES_KEY_TYPE __global const
+#define AES_SRC_TYPE MAYBE_CONSTANT
+#include "opencl_aes.h"
+
+typedef struct {
+	uint cracked;
+} ab_out;
+
+typedef struct {
+	pbkdf2_salt pbkdf2;
+	int masterkey_blob_length;
+	uchar iv[16];
+	uchar masterkey_blob[MAX_MASTERKEYBLOB_LEN];
+} ab_salt;
+
+inline int ab_decrypt(__global uchar *key, MAYBE_CONSTANT ab_salt *salt)
+{
+	uchar out[MAX_MASTERKEYBLOB_LEN];
+	const int length = salt->masterkey_blob_length;
+	uchar aiv[16];
+	AES_KEY akey;
+	int pad_byte;
+
+	memcpy_macro(aiv, salt->iv, 16);
+	AES_set_decrypt_key(key, 256, &akey);
+	AES_cbc_decrypt(salt->masterkey_blob, out, length, &akey, aiv);
+
+	if (out[0] != 16)
+		return 0;
+
+	if (check_pkcs_pad(out, length, 16) < 0)
+		return 0;
+
+	pad_byte = out[length - 1];
+	if (pad_byte > 8)
+		return 1;
+
```

Comment:
Are you sure this will not result in false negatives?

Reply:
What I meant is: Are you sure there can't be a valid padding of 8 or less? Why is that?

Reply:
My bad, I missed reading the false *negatives* word correctly.

I have added the following comment to the code regarding this.

```
// The structure of masteykey_blob is fixed:
//   + masteykey_blob -> length_1 (byte) + IV (16 bytes) + length_2 + masteykey (32 bytes) + length_3 + checksum (32 bytes) => total of 83 bytes
//   + padding -> this data blob of 83 bytes gets 13 bytes of padding making the masteykey_blob_length = 96 bytes always
```

Reply:
This implies that padding is always 13 bytes in practice.

Reply:
Perfect!